Check feature flag default values at compile time

emw-apple opened the pull request:
#### 80b7b85c4493d6cdc616efd9b10e2ba1914a0b17
<pre>
Check feature flag default values at compile time
<a href="https://bugs.webkit.org/show_bug.cgi?id=258041">https://bugs.webkit.org/show_bug.cgi?id=258041</a>
rdar://104759188

Reviewed by Darin Adler.

Feature flags which have a constant default value (true or false, not
determined at runtime) can be checked statically to ensure that their
default value corresponds with their feature status&apos;s implied default.

First, the preferences generator emits a std::bool_constant for boolean
literals when generating WebKit&apos;s feature default values. It also emits
the feature&apos;s status as a constant type.

Then, a new template overload to APIFeature::create() accepts a feature
status constant and a std::bool_constant default value, and
static-asserts that the default values match.

Assertion failures contain a template instation note, so it&apos;s easy to
see which feature caused the failure:

    In file included from /Volumes/Data/OpenSource/WebKitBuild/Debug/DerivedSources/WebKit/WebPreferencesFeatures.cpp:29:
    In file included from /Volumes/Data/OpenSource/Source/WebKit/UIProcess/WebPreferences.h:28:
    /Volumes/Data/OpenSource/Source/WebKit/UIProcess/API/APIFeature.h:48:13: error: static assertion failed due to requirement &apos;!defaultValue&apos;: Feature&apos;s status implies it should be off by default
                static_assert(!defaultValue, &quot;Feature&apos;s status implies it should be off by default&quot;);
                ^             ~~~~~~~~~~~~~
    /Volumes/Data/OpenSource/WebKitBuild/Debug/DerivedSources/WebKit/WebPreferencesFeatures.cpp:688:23: note: in instantiation of function template specialization &apos;API::Feature::create&lt;API::FeatureStatus::Developer, true&gt;&apos; requested here
            API::Feature::create(&quot;Prefer Page Rendering Updates near 60fps&quot;&quot;&quot;_s, &quot;PreferPageRenderingUpdatesNear60FPSEnabled&quot;_s, API::FeatureConstant&lt;API::FeatureStatus::Developer&gt;{}, API::FeatureCategory::DOM, &quot;Prefer page rendering updates near 60 frames per second rather than using the display&apos;s refresh rate&quot;&quot;&quot;_s, DEFAULT_VALUE_FOR_PreferPageRenderingUpdatesNear60FPSEnabled, false),
                          ^
    1 error generated.

This mechanism is controlled by a ENABLE(FEATURE_DEFAULT_VALIDATION)
flag, which is currently only enabled on PLATFORM(MAC).

* Source/WTF/Scripts/GeneratePreferences.rb:
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml: Change the
  status of two features whose defaults are not aligned:
  - PreferPageRenderingUpdatesNear60FPSEnabled, developer -&gt; stable:
    It&apos;s on by default everywhere except visionOS.
  - MediaSourceInlinePaintingEnabled, unstable -&gt; stable: It&apos;s on for
    all Apple platforms since last year.
* Source/WebKit/Scripts/PreferencesTemplates/WebPreferencesDefinitions.h.erb:
* Source/WebKit/UIProcess/API/APIFeature.cpp:
(API::Feature::uncheckedCreate):
(API::Feature::create): Deleted.
* Source/WebKit/UIProcess/API/APIFeature.h:
* Source/WebKit/UIProcess/API/APIFeatureStatus.h:
(API::defaultValueForFeatureStatus):

Canonical link: <a href="https://commits.webkit.org/269988@main">https://commits.webkit.org/269988@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b55343d14e5b60c92fee110bfbb509eac9c0247c

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/24198 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/2307 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/25280 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/26330 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/22281 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/3934 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/24667 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/22738 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/24441 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/1829 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/20914 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/26919 "Built successfully") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/24369 "Passed tests") | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/1577 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/21833 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/28055 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/21043 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/22060 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/22130 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/25835 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/23496 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/1514 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/19176 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/30894 "Built successfully") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/2167 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/6779 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/5798 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/1920 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/30858 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/1878 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/6454 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

emw-apple commented on Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
```diff
@@ -4309,7 +4309,7 @@ MediaSourceEnabled:
 MediaSourceInlinePaintingEnabled:
   type: bool
-  status: unstable
+  status: stable
```
@jernoble: Could you confirm that this status change is accurate? It reflects that the feature is always enabled on Mac.

darinadler replied:
I think it’s risky to have a single status express the state of the feature, and have a road map to enforcing the relationship, when the status may be different per platform and even per OS version. The fact that `defaultValue` is conditional here is quite important!

emw-apple replied:
Indeed. Currently, macOS is the platform that the statuses are aligned to, because Mac Safari is the only browser that exposes feature status to web developers.

IMO, the next step will be making it possible for a feature's status to be conditional, so that they can accurately describe behavior on different platforms.

emw-apple commented on Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
```diff
@@ -5015,7 +5015,7 @@ PreferFasterClickOverDoubleTap:
 PreferPageRenderingUpdatesNear60FPSEnabled:
   type: bool
-  status: developer
+  status: stable
```
@smfr: Could you confirm that this status change is accurate? It reflects that the feature is always enabled on Mac.

darinadler replied:
The unusual platform here is visionOS. Again, saying this is "stable" in a platform-independent way may not be right for something we can’t turn on for visionOS. This is a strategic question, I think.

Maybe we’re saying that feature status is mostly about macOS?

darinadler commented on Source/WTF/wtf/PlatformEnable.h:
```diff
@@ -1006,3 +1006,9 @@
+#if !defined(ENABLE_FEATURE_DEFAULT_VALIDATION) \
+    && (PLATFORM(MAC) && __MAC_OS_X_VERSION_MIN_REQUIRED >= 140000)
+#define ENABLE_FEATURE_DEFAULT_VALIDATION 1
+#endif
```
I am assuming the reason here is that eventually we want to do this everywhere, but doing it slowly lets you make incremental progress. But that’s not at all clear. In my opinion, we need somewhere in the code where the "why" is explained. This internal self-checking is not much like the other features we enable, which typically are either web-exposed or end-user features.

emw-apple replied:
> I am assuming the reason here is that eventually we want to do this everywhere, but doing it slowly lets you make incremental progress.

Exactly. I'll leave a FIXME here explaining our desire to bring this to additional platforms.

darinadler commented on Source/WebKit/UIProcess/API/APIFeatureStatus.h:
```diff
@@ -45,6 +48,28 @@ enum class FeatureStatus : uint8_t {
+        return false;
+    case FeatureStatus::Embedder:
+    case FeatureStatus::Internal:
+        return { };
```
Why is it right to just let these features go either way? I’m not doubting you, but it’s not obvious so might need a brief comment.

emw-apple replied:
Added a comment. It's for two different reasons: Embedder features are intended to vary widely between platforms, so they don't have an implied default value.

Internal features are supposed to be off by default (as documented in `UnifiedWebPreferences.yaml`), but we have a number of existing internal features that break the rule. Since these are not user-visible, I don't think we need to correct them all now—we gain most of the benefits by having internal checking of the public statuses.

darinadler reviewed (approved):
Good approach. I am sure you’ll want feedback from others but I am good with this as is.

emw-apple reviewed (commented):
Thanks for the review, and sorry for letting this simmer for so long.

There's some renewed interest in getting this landed, but we'd like to pull the changes to individual features' statuses out to separate commits. I'll get those reviewed and landed first (in #19652 and #19653), and then finish by landing the generator changes.

emw-apple committed:
Check feature flag default values at compile time
https://bugs.webkit.org/show_bug.cgi?id=258041
rdar://104759188

Reviewed by Darin Adler.

Feature flags which have a constant default value (true or false, not
determined at runtime) can be checked statically to ensure that their
default value corresponds with their feature status's implied default.

First, the preferences generator emits a std::bool_constant for boolean
literals when generating WebKit's feature default values. It also emits
the feature's status as a constant type.

Then, a new template overload to APIFeature::create() accepts a feature
status constant and a std::bool_constant default value, and
static-asserts that the default values match.

Assertion failures contain a template instation note, so it's easy to
see which feature caused the failure:

    In file included from /Volumes/Data/OpenSource/WebKitBuild/Debug/DerivedSources/WebKit/WebPreferencesFeatures.cpp:29:
    In file included from /Volumes/Data/OpenSource/Source/WebKit/UIProcess/WebPreferences.h:28:
    /Volumes/Data/OpenSource/Source/WebKit/UIProcess/API/APIFeature.h:48:13: error: static assertion failed due to requirement '!defaultValue': Feature's status implies it should be off by default
                static_assert(!defaultValue, "Feature's status implies it should be off by default");
                ^             ~~~~~~~~~~~~~
    /Volumes/Data/OpenSource/WebKitBuild/Debug/DerivedSources/WebKit/WebPreferencesFeatures.cpp:688:23: note: in instantiation of function template specialization 'API::Feature::create<API::FeatureStatus::Developer, true>' requested here
            API::Feature::create("Prefer Page Rendering Updates near 60fps"""_s, "PreferPageRenderingUpdatesNear60FPSEnabled"_s, API::FeatureConstant<API::FeatureStatus::Developer>{}, API::FeatureCategory::DOM, "Prefer page rendering updates near 60 frames per second rather than using the display's refresh rate"""_s, DEFAULT_VALUE_FOR_PreferPageRenderingUpdatesNear60FPSEnabled, false),
                          ^
    1 error generated.

This mechanism is controlled by a ENABLE(FEATURE_DEFAULT_VALIDATION)
flag, which is currently only enabled on PLATFORM(MAC).

* Source/WTF/Scripts/GeneratePreferences.rb:
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml: Change the
  status of two features whose defaults are not aligned:
  - PreferPageRenderingUpdatesNear60FPSEnabled, developer -> stable:
    It's on by default everywhere except visionOS.
  - MediaSourceInlinePaintingEnabled, unstable -> stable: It's on for
    all Apple platforms since last year.
* Source/WebKit/Scripts/PreferencesTemplates/WebPreferencesDefinitions.h.erb:
* Source/WebKit/UIProcess/API/APIFeature.cpp:
(API::Feature::uncheckedCreate):
(API::Feature::create): Deleted.
* Source/WebKit/UIProcess/API/APIFeature.h:
* Source/WebKit/UIProcess/API/APIFeatureStatus.h:
(API::defaultValueForFeatureStatus):

Canonical link: https://commits.webkit.org/269988@main
